thunder/torch/__init__.py

```diff
@@ -186,6 +186,92 @@ def is_cuda(a: TensorLike, /) -> bool:
 
 register_method("size", size)
 
+_torch_dtype_to_old_torch_typestring_map = {
+    torch.float32: "FloatTensor",
+    torch.float64: "DoubleTensor",
+    torch.float16: "HalfTensor",
+    torch.bfloat16: "BFloat16Tensor",
+    torch.uint8: "ByteTensor",
+    torch.int8: "CharTensor",
+    torch.int16: "ShortTensor",
+    torch.int32: "IntTensor",
+    torch.long: "LongTensor",
+    torch.bool: "BoolTensor",
+}
+
+_old_torch_typestring_to_torch_dtype_map = {v: k for k, v in _torch_dtype_to_old_torch_typestring_map.items()}
+
+
+def _device_and_dtype_to_old_torch_typestring(device: DeviceLike, dtype: dtypeLike):
+    torch_dtype = to_torch_dtype(dtype)
+    dtype_str = _torch_dtype_to_old_torch_typestring_map.get(torch_dtype)
+    devicetype_str: str = ""
+    if device.devicetype is not devices.DeviceType.CPU:
+        devicetype_str = f"{devices.devicetype_string(device.devicetype)}."
+    return f"torch.{devicetype_str}{dtype_str}"
+
+
+def _old_torch_typestring_to_devicetype_and_dtype(typestring: str) -> tuple[DeviceLike, dtypeLike]:
+
+    # Two cases:
+    #    - torch.DtypeTensor
+    #    - torch.device.DtypeTensor
+
+    _, *dev_and_dtype = typestring.split(".")
+
+    if len(dev_and_dtype) == 1:
+        (dtype_str,) = dev_and_dtype
+        return "cpu", _old_torch_typestring_to_torch_dtype_map[dtype_str]
+
+    if len(dev_and_dtype) == 2:
+        dtype_str, devicetype_str = dev_and_dtype
+        return dtype_str, _old_torch_typestring_to_torch_dtype_map[devicetype_str]
+
+    # Assertion error -- expected the string to split into one or two elements
+    utils.check(
+        False,
+        lambda: f"type(): dtype format does not match torch.Tensor dtype nor old torch typestring format",
+        exception_type=ValueError,
+    )
+
+
+@torchsymbol(torch.Tensor.type, is_method=True)
+def type(
+    a: TensorLike, /, dtype: None | str | dtypeLike = None, non_blocking: bool = False, **kwargs
+) -> str | TensorLike:
+    utils.check(
+        not non_blocking,
+        lambda: f"type(): `non_blocking==True` is currently not supported.",
+        exception_type=NotImplementedError,
+    )
+
+    if dtype is None:
+        # returns the type of the input tensor in string
+        return _device_and_dtype_to_old_torch_typestring(a.device, a.dtype)
+
+    if isinstance(dtype, str):
+        devtype, dtype = _old_torch_typestring_to_devicetype_and_dtype(dtype)
+
+        # below if-statement handles the following case that you have mentioned:
+        # Follow-up question: what if the old string says "CUDA" but the tensor is on CUDA device 1, will a tensor
+        # on CUDA device 0 be created?
+
+        if devtype == a.device.type:
+            dev = a.device
+        elif devtype == "cpu":
+            dev = devices.DeviceType.CPU
+        else:
+            dev = devices.DeviceType.CUDA
```

Comment:
Let's not assume that the fallthrough is `CUDA`. I think this can work like this:

```
if devtype == a.device.type:
  # This handles the case that a tensor is already on a CUDA device, and the device type string is CUDA. In this
  # case the tensor remains on its current device.
  dev = a.device
else:
  dev = device_from_string(devtype)
```

Reply:
The issue with this is that, `device_from_string` requires a device number in CUDA. However, the weird PyTorch typestring does not include that.

Reply:
Let me do something like this to avoid such assumption.
```python
        if devtype == a.device.type:
            dev = a.device
        elif devtype == "cpu":
            dev = devices.DeviceType.CPU
        elif devtype == "cuda":
            dev = devices.DeviceType.CUDA
        else:
            raise ValueError(f"type(): unrecognized torch typestring {dtype}")
```

Reply:
I appreciate your detailed looking at this function, including how it will show errors. I agree with you the current error message for that case is not very good.

However, if possible, I think we should improve the `_old_torch_typestring_to_devicetype_and_dtype` function to throw a better error message when given something ridiculous. That way any future callers of the function will also throw better error messages, so the code will be more maintainable. Otherwise each caller would have to include logic like the above.

So may before returning from `_old_torch_typestring_to_devicetype_and_dtype`, it should check that the devicetype string is inferred to be `cpu` or `cuda`?

Reply:
I see what you mean. Thanks for the detailed explanation. Let me add that check in `_old_torch_typestring_to_devicetype_and_dtype`!